### PR TITLE
Use S3 Select for flow result exports

### PR DESCRIPTION
**Pull request by rowanseymour**
Once it we're sure this works, can also use for message and contact exports

### CODE REVIEW

File: temba/tests/s3.py [hunk 1]
```diff
@@ -74,7 +80,72 @@ def select_object_content(self, Bucket, Key, **kwargs):
...
-            records.append(json.loads(line.decode("utf-8")))
+            record = json.loads(line.decode("utf-8"))
+
+            if not Expression or select_matches(Expression, record):
+                records.append(record)
 
         return {"Payload": MockEventStream(records)}
+
+
+def select_matches(expression: str, record: dict) -> bool:
```
Comment (rowanseymour) [inline temba/tests/s3.py:91]: SQL engine in 60 lines.. watch out Postgres 😎

File: temba/archives/models.py [hunk 2]
```diff
@@ -144,53 +143,67 @@ def _get_covering_period(cls, org, archive_type: str, after: datetime = None, be
...
-            stream = gzip.GzipFile(fileobj=s3_obj["Body"])
-
-            while True:
-                line = stream.readline()
-                if not line:
-                    break
+        s3_client = s3.client()
 
-                yield json.loads(line.decode("utf-8"))
+        def generator():
```
Comment (rowanseymour) [inline temba/archives/models.py:183]: learnt something new about generators - if the method itself is a generator, i.e. has `yield` statements.. you can only call it once. By nesting here we allow `Archive.iter_records` to be called more than once and return different generators.

**Review by ericnewcomer — APPROVED** [review]
This looks awesome, super curious how this performs on giant s3 files. Did you test some queries out on the command line against some 800MB files?

**Review by norkans7 — APPROVED** [review]

**Comment by rowanseymour** [comment]
All depends on how much data you're leaving on the server but I've tested it before and its night and day if you're only fetching a few rows from an 800GB file

**Pull request merged**